sw/airborne/modules/dragspeed/dragspeed.c

```diff
@@ -0,0 +1,228 @@
+#endif
+
+#ifndef DRAGSPEED_COEFF_Y
+#define DRAGSPEED_COEFF_Y 1.0 /// Drag coefficient (mu/m) of the linear drag model along y axis, where m*a = v*mu
+#endif
+
+#ifndef DRAGSPEED_ZERO_X
+#define DRAGSPEED_ZERO_X 0.0 /// Accelerometer reading (x axis) when stationary [m/s^2]
+#endif
+
+#ifndef DRAGSPEED_ZERO_Y
+#define DRAGSPEED_ZERO_Y 0.0 /// Accelerometer reading (y axis) when stationary [m/s^2]
+#endif
+
+#ifndef DRAGSPEED_R
+#define DRAGSPEED_R 0.25 /// Velocity measurement noise variance [(m/s)^2]
+#endif
+
+#ifndef DRAGSPEED_FILTER
+#define DRAGSPEED_FILTER 0.8 /// First-order low-pass filter strength [0..1]. Pretty high default value as accelero's tend to be noisy.
+#endif
+
+struct dragspeed_t dragspeed;
+
+static abi_event accel_ev;
+static void accel_cb(
+		uint8_t sender_id,
+		uint32_t stamp,
+		struct Int32Vect3 *accel);
+
+static void send_dragspeed(struct transport_tx *trans, struct link_device *dev);
+
+static void calibrate_coeff(struct Int32Vect3 *accel);
+static void calibrate_zero(struct Int32Vect3 *accel);
+
+void dragspeed_init(void) {
+	// Set initial values
+	dragspeed.coeff.x = DRAGSPEED_COEFF_X;
+	dragspeed.coeff.y = DRAGSPEED_COEFF_Y;
+	dragspeed.filter = DRAGSPEED_FILTER;
+	// Register callbacks
+	register_periodic_telemetry(DefaultPeriodic, PPRZ_MSG_ID_DRAGSPEED,
+			send_dragspeed);
+	AbiBindMsgIMU_ACCEL_INT32(DRAGSPEED_ACCEL_ID, &accel_ev, accel_cb);
+}
+
+bool dragspeed_calibrate_coeff(void) {
+	dragspeed.do_calibrate_coeff = 1;
+	return FALSE;
+}
+
+bool dragspeed_calibrate_zero(void) {
+	dragspeed.do_calibrate_zero = 1;
+	return FALSE;
+}
+
+bool dragspeed_is_calibrating(void) {
+	return dragspeed.do_calibrate_coeff || dragspeed.do_calibrate_zero;
+}
+
+static void accel_cb(
+		uint8_t sender_id __attribute__((unused)),
+		uint32_t stamp,
+		struct Int32Vect3 *accel) {
+	// Estimate current velocity
+	float vx = -(ACCEL_FLOAT_OF_BFP(accel->x) - dragspeed.zero.x)
+			/ dragspeed.coeff.x;
+	float vy = -(ACCEL_FLOAT_OF_BFP(accel->y) - dragspeed.zero.y)
+			/ dragspeed.coeff.y;
+	// Simple low-pass filter
+	dragspeed.vel.x += (1 - dragspeed.filter) * (vx - dragspeed.vel.x);
+	dragspeed.vel.y += (1 - dragspeed.filter) * (vy - dragspeed.vel.y);
+	// Send as ABI VELOCITY_ESTIMATE message
+	// Note: set VEL_DRAGSPEED_ID to ABI_DISABLE to disable
+#if VEL_DRAGSPEED_ID
+	if (!dragspeed.do_calibrate_coeff && !dragspeed.do_calibrate_zero) {
+		AbiSendMsgVELOCITY_ESTIMATE(VEL_DRAGSPEED_ID, stamp,
+				dragspeed.vel.x, dragspeed.vel.y, 0, DRAGSPEED_R);
+	}
+#endif
+	// Perform calibration if required
+	calibrate_coeff(accel);
+	calibrate_zero(accel);
+}
+
+/**
+ * Calibrate drag coefficient by comparing accelerometer measurements to INS
+ * velocities.
+ * Should be performed with VEL_DRAGSPEED_ID set to ABI_DISABLE, but for safety
+ * the ABI messages are also disabled automatically when calibration is active.
+ *
+ * This routine assumes that the accelerometers have been zeroed beforehand.
```

Comment:
Why assume, if this important for the operation of the calibration routine you could add a flag to the dragspeed struct to include a flag that indicates if the accel has been zero'ed?